### DISCUSSION TITLE
How to debug Error: write EPIPE during build for a 1000+ pages site

### DISCUSSION THREAD

**Discussion by gvocale** [gatsbyjs/gatsby#34043] [Help]
### Preliminary Checks

- [X] This issue is not a duplicate. Before opening a new issue, please search existing issues: https://github.com/gatsbyjs/gatsby/issues
- [X] This issue is not a question, feature request, RFC, or anything other than a bug report directly related to Gatsby. Please post those things in GitHub Discussions: https://github.com/gatsbyjs/gatsby/discussions

### Description

During `yarn build`, the following `Error: write EPIPE` happens. What does this error mean? How can I debug it?

My site has around 1100 pages.

Thanks

### Reproduction Link

https://github.com/

### Steps to Reproduce

1. `yarn develop`
2. Wait wait wait...
3. Error happens Error: write EPIPE
...

### Expected Result

Build to work

### Actual Result

```tsx
success run queries in workers - 84.981s - 1613/1613 18.98/s
⠋ Merge worker state
/site/node_modules/yoga-layout-prebuilt/yoga-layout/build/Release/nbind.js:53
        throw ex;
        ^

Error: write EPIPE
    at ChildProcess.target._send (node:internal/child_process:849:20)
    at ChildProcess.target.send (node:internal/child_process:722:19)
    at WorkerPool.sendMessage (/site/node_modules/gatsby-worker/dist/index.js:290:19)
    at /site/node_modules/gatsby/src/utils/jobs/worker-messaging.ts:22:22
Emitted 'error' event on ChildProcess instance at:
    at node:internal/child_process:853:39
    at processTicksAndRejections (node:internal/process/task_queues:78:11) {
  errno: -32,
  code: 'EPIPE',
  syscall: 'write'
}
error Command failed with exit code 7.
```

### Environment

```shell
System:
    OS: macOS 12.0.1
    CPU: (16) x64 Intel(R) Core(TM) i9-9880H CPU @ 2.30GHz
    Shell: 5.8 - /bin/zsh
  Binaries:
    Node: 16.6.0 - /var/folders/k0/zk60y_zj5v518rw_4hp4k0840000gn/T/yarn--1637333410517-0.2146436190636809/node
    Yarn: 1.22.11 - /var/folders/k0/zk60y_zj5v518rw_4hp4k0840000gn/T/yarn--1637333410517-0.2146436190636809/yarn
    npm: 7.19.1 - ~/.nvm/versions/node/v16.6.0/bin/npm
  Languages:
    Python: 2.7.18 - /usr/bin/python
  npmPackages:
    gatsby: ^4.1.4 => 4.1.6
    gatsby-plugin-disqus: ^1.2.3 => 1.2.3
    gatsby-plugin-google-analytics: ^4.1.0 => 4.1.0
    gatsby-plugin-image: ^2.1.2 => 2.1.3
    gatsby-plugin-manifest: ^4.1.3 => 4.1.4
    gatsby-plugin-react-helmet: ^5.1.0 => 5.1.0
    gatsby-plugin-sass: ^5.1.1 => 5.1.1
    gatsby-plugin-sharp: ^4.1.3 => 4.1.4
    gatsby-plugin-sitemap: ^5.1.1 => 5.1.1
    gatsby-plugin-theme-ui: ^0.12.0 => 0.12.0
    gatsby-plugin-typescript: ^4.1.2 => 4.1.3
    gatsby-source-filesystem: ^4.1.2 => 4.1.3
    gatsby-transformer-json: ^4.1.0 => 4.1.0
    gatsby-transformer-remark: ^5.1.3 => 5.1.4
    gatsby-transformer-sharp: ^4.1.0 => 4.1.0


✨  Done in 2.27s.
```

**Comment by KyleAMathews** [comment]
Generally this means a worker process was killed by your OS as the machine was running out of memory. `Error: write EPIPE` means the parent Gatsby processed was unable to write to the child anymore.

**Comment by gvocale** [comment]
Thank you, very helpful! I will try to run my build process with:
```
"build": "export NODE_OPTIONS=--max_old_space_size=8192 && gatsby build",
```
And see if it helps

**Reply by gvocale** [reply]
I have tried increasing the memory, but that didn't seem to help....even something like `export NODE_OPTIONS=--max_old_space_size=32192 && gatsby build` would still end in a `Error: write EPIPE`

**Reply by rburgst** [reply]
same problem here, the problem was introduced when activating Parallel Query running which as of gatsby 4 can no longer be disabled: https://github.com/gatsbyjs/gatsby/discussions/32389#discussioncomment-1945226

**Reply by witcradg** [reply]
Ugh.. Not looking forward to rolling back to Gatsby 3, but I can't do production releases any more. This was not a problem with Gatsby 3.

**Comment by realharry** [comment]
I am having the same issue. I have been having intermittent issues starting from > 500 pages. Now, it's over 1000 pages, and it does not build at all. this is a critical issue for me. is there any workaround without having to downgrade gatsby (and all dependent package)?

**Reply by rburgst** [reply]
We also ran into these problems and unfortunately gatsby cant seem to get a handle on those problems, and actually with the introduction of even newer features (IMAGE_CDN I am looking at you), they are introducing even more problems (https://github.com/gatsbyjs/gatsby/issues/35151). At any rate here are some tips that might help here

https://www.gatsbyjs.com/docs/how-to/performance/resolving-out-of-memory-issues/

I generally feel that with newer gatsby versions the RAM requirements have risen through the roof and very often if you are running out of RAM then gatsby fails silently with exactly this error message.

Especially if you are hosting your own build runners this might be causing problems (we had to bump the mem that we gave to our runners *considerably* to get the build to go through).

One thing that might help here is moving your builds to gatsby cloud (I am not affiliated with them in any form), there, their runners seem to be well spec'ed and you might have more luck getting it to work.

**Reply by realharry** [reply]
Hi @rburgst Thanks very much for the tip. I tried a few different things suggested by the doc, but no luck so far. Looks like I'll have to add more RAM. I'm just concerned that, even if it solves this issue for now, I'll likely run into the same issue at some point in the future when I add more pages and more images. I feel like we'll need more fundamental solution to this issue. Thanks, and have a wonderful day/night. :)

**Comment by witcradg** [comment]
@rburgst @realharry 
I'm not sure what you're using for NODE, but I found issues with NODE v17...
Also, I can't testify to the effiectiveness of the NODE_OPTIONS below, but this worked for me to eliminate the EPIPE errors.
NODE_OPTIONS --max-old-space-size=8192
NODE_VERSION 16.13.1

Other unrelated but useful settings I use:  
NETLIFY_SKIP_GATSBY_FUNCTIONS   true
NPM_FLAGS --legacy-peer-deps

**Reply by witcradg** [reply]
...and yes, I'm running more than 1000 pages

**Reply by rburgst** [reply]
note that its not about the number of pages. Its more about the number and size of images as typically the sharp processes die.

**Reply by realharry** [reply]
@rburgst @witcradg I haven't had much luck with --max-old-space-size NODE_OPTIONS. I tried various combinations, either passing it as an argument to node (e.g., "node --max-old-space-size=8000 gatsby build") or using as an env variables (e.g., "export NODE_OPTIONS=--max-old-space-size=8000" and then running gatsby build, etc. etc.) and all possible combinations that I could think of. If I run with process-top, it always dies before the memory reaches 2G, more like between 1 and 2G. It makes me think that the node options is not effective in my case. (I'm using node 16.14.1.) 

Yes, i have a suspicion that it is the number/sizes of images, not the page number in my case. It's the "Running gatsby-plugin-sharp.IMAGE_PROCESSING jobs" that seems to be having an issue. for references, i have ~1000 jpeg images with 1280x720 resolutions.

Adding more physical memory (8G) to my machine is a bit of a hassle at this point, and I'm still trying to find a workaround before I go that route...

**Reply by witcradg** [reply]
@realharry For me the key was downgrading NODE from 17 to 16.

**Reply by realharry** [reply]
@witcradg Thanks!! Let me try different versions of node. 👍

**Comment by komanton** [comment]
I decreased the amount of CPUs of the my build docker container.
And, the issue was fixed (without `export NODE_OPTIONS=--max_old_space_size=32192`).

Probably, you need to find your balanced value of CPU and free RAM (Also, you can use swap). 
My configuration for docker desktop: 4 CPUs, 7.8 Gb RAM, 1Gb Swap.
